jack-test/test/java/com/rapleaf/jack/transaction/TestDbTransactorImpl.java

```diff
@@ -38,6 +44,43 @@ public void cleanup() throws Exception {
     executorService = null;
   }
 
+  @Test
+  public void testConnectionsReturnedOnRollbackFailure() {
+    MysqlDatabaseConnection realConnection = new MysqlDatabaseConnection("database1");
+    BaseDatabaseConnection mBaseDatabaseConnection = Mockito.spy(realConnection);
+    IDatabase1 database1 =
+        new Database1Impl(
+            mBaseDatabaseConnection,
+            new DatabasesImpl(mBaseDatabaseConnection),
+            new NoOpAction()
+        );
+
+    Mockito.doAnswer(invocationOnMock1 -> {
+      throw new SQLRecoverableException("Commit failure");
+    }).when(mBaseDatabaseConnection).commit();
+
+    Mockito.doAnswer(invocationOnMock -> {
+      throw new SQLRecoverableException("Rollback failure");
+    }).when(mBaseDatabaseConnection).rollback();
+
+    DbPoolManager<IDatabase1> dbPoolManager = Mockito.spy(new DbPoolManager<>(() -> database1, 1,
+        1, 100, Integer.MAX_VALUE, false
+    ));
+    TransactorImpl<IDatabase1> transactor = new TransactorImpl<>(dbPoolManager, false);
+
+    try {
+      transactor.queryAsTransaction(db -> db.users().find(1));
+      fail();
+    } catch (Exception e) {
+      // This should fail
+    }
+
+    Mockito.verify(dbPoolManager, Mockito.times(1)).returnConnection(Mockito.any(IDatabase1.class));
+
+    // This simply needs to not throw, demonstrating that the connection pool is not exhausted.
+    transactor.query(db -> db.users().find(1));
```

Comment:
I hate a bajillion magic numbers. 

Can we perhaps assign `1` to some variable at the top of this test and then re-use it so its clear what the relationship of the values are?

Reply:
Ah. Yeah, I can assign it to a variable and leave a comment. It basically has no relation. I just need to query for something. It doesn't matter that it happens to be the same id in both the queries.